pie/trainer.py

```diff
@@ -215,6 +224,74 @@ def __init__(self, settings, model, dataset, num_instances):
+        reader = Reader(settings, settings.input_path)
+        tasks = reader.check_tasks(expected=None)
+
+        # label encoder
+        label_encoder: MultiLabelEncoder = MultiLabelEncoder.from_settings(settings, tasks=tasks)
+        label_encoder.fit_reader(reader)
+
+        trainset = Dataset(settings, reader, label_encoder)
+
+        devset = None
+        if settings.dev_path:
+            devset = Dataset(settings, Reader(settings, settings.dev_path), label_encoder)
+        else:
+            logging.warning("No devset: cannot monitor/optimize training")
+
+        # model
+        model: SimpleModel = SimpleModel(
+            label_encoder, settings.tasks,
+            settings.wemb_dim, settings.cemb_dim, settings.hidden_size,
+            settings.num_layers, cell=settings.cell,
+            # dropout
+            dropout=settings.dropout, word_dropout=settings.word_dropout,
+            # word embeddings
+            merge_type=settings.merge_type, cemb_type=settings.cemb_type,
+            cemb_layers=settings.cemb_layers, custom_cemb_cell=settings.custom_cemb_cell,
+            # lm joint loss
+            include_lm=settings.include_lm, lm_shared_softmax=settings.lm_shared_softmax,
+            # decoder
+            scorer=settings.scorer, linear_layers=settings.linear_layers
+        )
+
+        # pretrain(/load pretrained) embeddings
+        if model.wemb is not None:
+            if settings.pretrain_embeddings:
+                print("Pretraining word embeddings")
+                wemb_reader = Reader(
+                    settings, settings.input_path, settings.dev_path, settings.test_path)
+                weight = get_pretrained_embeddings(
+                    wemb_reader, label_encoder, size=settings.wemb_dim,
+                    window=5, negative=5, min_count=1)
+                model.wemb.weight.data = torch.tensor(weight, dtype=torch.float32)
+
+            elif settings.load_pretrained_embeddings:
+                print("Loading pretrained embeddings")
+                if not os.path.isfile(settings.load_pretrained_embeddings):
+                    print("Couldn't find pretrained eembeddings in: {}".format(
+                        settings.load_pretrained_embeddings))
+                pie.initialization.init_pretrained_embeddings(
+                    settings.load_pretrained_embeddings, label_encoder.word, model.wemb)
+
+        # load pretrained weights
+        if settings.load_pretrained_encoder:
+            model.init_from_encoder(pie.pretrain_encoder.Encoder.load(settings.load_pretrained_encoder))
+
+        # freeze embeddings
+        if settings.freeze_embeddings:
+            model.wemb.weight.requires_grad = False
+
+        model.to(settings.device)
+
+        trainer = cls(settings, model, trainset, reader.get_nsents())
+
+        #"Trainer", SimpleModel, Dataset, Optional[Dataset], MultiLabelEncoder, Reader
+        return trainer, model, trainset, devset, label_encoder, reader
```

Comment:
Trainer class already has instance variables to model and trainset. And label_encoder is always part of the model. The last reader is also part of trainset, so not sure we need this line.

Reply:
I need to look for the devset at least, I somehow missed some stuff.